pkg/kv/kvserver/kvserverpb/raft.proto

```diff
@@ -100,6 +100,35 @@ message RaftMessageRequest {
 
 message RaftMessageRequestBatch {
   repeated RaftMessageRequest requests = 1 [(gogoproto.nullable) = false];
+
+  // StoreIDs identifies all the stores on the client node. It's populated on
+  // the first RaftMessageRequestBatch sent along MultiRaft.RaftMessageBatch
+  // gRPC stream identifying at least one store, and the populated once more
+  // if any additional stores have been initialized[^1]. This data is used by
+  // the kvflowcontrol machinery to track the exact set of stores on the client
+  // node. It uses this information to react to the gRPC streams breaking. Since
+  // these streams are used to piggy information about which log entries were
+  // admitted below raft[^2] in order for the server-side to free up flow
+  // tokens, if the stream breaks we possibly risk leaking these tokens. So
+  // when these streams break, we use information about the client's stores to
+  // release all held tokens[^3].
+  //
+  // [^1]: This two-step process is because of how and when we allocate
+  //       StoreIDs. Ignoring nodes that are bootstrapping the cluster (which
+  //       just picks the initial set of StoreIDs -- see
+  //       pkg/server.bootstrapCluster), whenever a new node is added, it's
+  //       assigned a node ID and store ID by an existing node in CRDB (see
+  //       kvpb.JoinNodeResponse). Subsequent store IDs, for multi-store nodes,
+  //       are generated by the joining node by incrementing a sequence ID
+  //       generator (see pkg/server.(*Node).initializeAdditionalStores). All of
+  //       which is to say that the very first time we issue a
+  //       RaftMessageRequestBatch, we might not have all the StoreIDs. But we
+  //       will very shortly after, and certainly before and replicas get
+  //       allocated to the additional store.
+  // [^2]: See kvflowcontrolpb.AdmittedRaftLogEntries and its use in
+  //       RaftMessageRequest.
+  // [^3]: See I1 from kvflowcontrol/doc.go.
+  repeated uint64 store_ids = 2 [(gogoproto.customname) = "StoreIDs", (gogoproto.casttype) = "github.com/cockroachdb/cockroach/pkg/roachpb.StoreID"];
```

Comment:
- We don't always have the `FromReplica`/`ToReplica` fields populated, because we're not always piggybacking token returns. We have a fallback mechanism in RaftTransport where we fire off kvflowcontrolpb.AdmittedRaftLogEntries if there haven't been any raft messages to piggy back on top of within kvadmission.flow_token.dispatch_interval (1s).
- That said, we _could_ construct this set of client-side stores on the server by intercepting every FromReplica.StoreID. If admission is deferred for some range's writes, presumably we'd have seen at least one raft message back from the follower replica, before the subsequent message that's either piggybacked (and containing the store ID itself) or not (in which case we already know the client's store ID). But we'd be intercepting this field on every raft message, instead of at most twice at stream creation time. I preferred the explicitness and the at-most-two-ness, it wasn't so involved.

I don't think the scheme above would be incorrect (== be possible to leak flow tokens). Consider the following:
- r1/1/n1/s1 (rangeID, replicaID, nodeID, storeID) deducts flow tokens for a write that's already replicated on r1/2/n2/s2.
- Because it knows about r1/2/n2/s2, i.e. it's added it part of the raft group previously, it's previously received some raft messages from r1/2/n2/s2 over the raft transport. Messages that identify that it's s2 on the other end through the FromReplica field. 
  - Say the raft transport stream's never broken up until this point. On any subsequent break we'll know to release tokens deducted for s2, i.e. the tokens we deducted for r1/2/n2/s2.
  - Say the raft transport has broken in the past, and had since re-established. We had added r1/2/n2/s2 when the transport was originally connected. Given that we deducted flow tokens for r1/2/n2/s2, we must seen raftTransportForFlowControl.isConnectedTo(s2) == true. So we've gotten _another_ raft message from some replica on n2/s2 over the raft transport stream from n2->n1 that notes that s2 is present on the other end. As before, on any subsequent break we'll know to release tokens deducted for s2, i.e. the tokens we deducted for r1/2/n2/s2.